ios/Video/RCTVideoPersistenceController.swift

```diff
@@ -0,0 +1,446 @@
+
+
+        // Create the AVAssetDownloadURLSession using the configuration.
+        assetDownloadURLSession =
+            AVAssetDownloadURLSession(configuration: backgroundConfiguration,
+                                      assetDownloadDelegate: self, delegateQueue: OperationQueue.main)
+        
+        
+        restorePersistenceManager()
+    }
+    
+    func restorePersistenceManager() {
+        guard !didRestorePersistenceController else { return }
+        
+        didRestorePersistenceController = true
+        
+        // Restore all tasks associated with the assetDownloadURLSession
+        assetDownloadURLSession.getAllTasks { tasksArray in
+            for task in tasksArray {
+                guard let assetDownloadTask = task as? AVAggregateAssetDownloadTask, let id = task.taskDescription else { break }
+                
+                // let urlAsset = assetDownloadTask.urlAsset
+
+                self.tasks[id] = assetDownloadTask
+            }
+            
+            // Restore all saved assets data from UserDefaults
+            guard let jsonString = self.userDefaults.value(
+                forKey: AssetPersistenceController.downloadedHlsDataKey
+            ) as? Data else {
+                return
+            }
+            
+            do {
+                let decoder = JSONDecoder()
+                let hlsAssets = try decoder.decode(
+                    [String: HLSAssetData].self,
+                    from: jsonString
+                )
+                for (id, hlsAssetData) in hlsAssets {
+                    let asset = HLSAsset(
+                        id: hlsAssetData.id,
+                        hlsURL: hlsAssetData.hlsURL,
+                        bookmark: hlsAssetData.bookmark,
+                        status: HLSAsset.DownloadState.init(
+                            rawValue: hlsAssetData.status
+                        )!,
+                        size: hlsAssetData.size
+                    )
+                    
+                    // Check if asset has corresponding AVURLAsset
+                    // (either from hlsURL or bookmark). This also makes sure
+                    // that downloaded asset's files were not deleted
+                    guard let urlAsset = asset.getURLAsset() else {
+                        continue
+                    }
+                   
+                    self.assets[id] = asset
+                    if (
+                        asset.status == HLSAsset.DownloadState.PENDING &&
+                        self.tasks[asset.id] == nil
+                    ) {
+                        // Download task was not recovered - delete leftover
+                        // data and recreate the task
+                        do {
+                            try FileManager.default.removeItem(at: urlAsset.url)
+                        } catch {
+                            print("Couldn't delete leftover data: \(error)")
```

Comment:
nah, this can happen if user started downloads, then force closed the app. Once the app restarts we want to resume downloads even if the download tasks couldn't be recovered